scheduler/remotesync.go

```diff
@@ -14,180 +14,114 @@
 }
 
-func (s *scheduler) monitorRemote(shutdown <-chan interface{}, remote, local client.Connection) {
-	var done <-chan interface{}
-	_shutdown := make(chan interface{})
-	cancel := make(chan struct{})
-
-	defer func() {
-		close(_shutdown)
-		close(cancel)
-		select {
-		case <-done:
-		}
-	}()
-
-	// monitor the leader's realm
-	ch := zzk.MonitorRealm(_shutdown, remote, "/scheduler")
-	for {
-		select {
-		case realm := <-ch:
-			switch realm {
-			case "":
-				// empty realm means something bad happened; exit
-				return
-			case s.realm:
-				// remote realm is the same as local realm; disconnect
-				// from the master until this changes
-				if done != nil {
-					cancel <- struct{}{}
-					<-done
-					done = nil
-				}
-			default:
-				// start remote synchonization if not yet started
-				if done == nil {
-					done = s.startRemote(cancel, remote, local)
-				}
-			}
-		case <-done:
-			// synchronization failed; shutdown
-			return
-		case <-shutdown:
-			// receieved signal to shutdown
-			return
-		}
-	}
+// RemoteSyncInterface is the endpoint where the data is synced
+type RemoteSyncInterface interface {
+	// GetServices returns all the services within a pool
+	GetServices(poolID string) ([]service.Service, error)
+	// AddService creates  a new service
+	AddService(svc *service.Service) error
+	// UpdateService updates an existing service
+	UpdateService(svc *service.Service) error
+	// DeleteService deletes a service
+	RemoveService(serviceID string) error
 }
 
-func (s *scheduler) remoteSync(shutdown <-chan interface{}, local client.Connection) {
-	for {
-		select {
-		case remote := <-zzk.Connect("/", zzk.GetRemoteConnection):
-			if remote != nil {
-				s.monitorRemote(shutdown, remote, local)
-			}
-		case <-shutdown:
-			return
-		}
-
-		select {
-		case <-shutdown:
-			return
-		default:
-			// probably lost connection to the remote; try again
-		}
-	}
+// RemoteServiceSync performs service synchronizations for services that
+// originate in remote pools.
+type RemoteServiceSync struct {
+	ds     RemoteSyncDatastore
+	iface  RemoteSyncInterface
+	poolID string
 }
 
-func (s *scheduler) getPoolSynchronizer() zzk.Listener {
-	poolSync := zkservice.NewPoolSynchronizer(s, zzk.GetRemoteConnection)
-
-	// Add the host listener
-	poolSync.AddListener(func(id string) zzk.Listener {
-		return zkservice.NewHostSynchronizer(s, id)
-	})
+// Purge looks up the remote data and performs the synchronization.
+// Implements utils.TTL
+func (sync *RemoteServiceSync) Purge(age time.Duration) (time.Duration, error) {
+	// look up the services
+	ids, svcs, err := sync.ds.GetUpdatedServices(sync.poolID)
+	if err != nil {
+		glog.Errorf("Could not look up remote services for pool %s: %s", sync.poolID, err)
+		return 0, err
+	}
 
-	// Add the service listener
-	poolSync.AddListener(func(id string) zzk.Listener {
-		return zkservice.NewServiceSynchronizer(s, id)
-	})
+	// build the datamap
+	datamap := make(map[string]interface{})
```

Comment:
I know utils.Sync takes a map[string]interface{}, but does this need to be generic? Will Go not compile this if it's a map[string]*service.Service?